runtime/src/mm/cpp/ThreadState.cpp

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+
+#include "ThreadData.hpp"
+#include "ThreadState.hpp"
+
+using namespace kotlin;
+
+namespace {
+
+// TODO: Can we make it constexpr? Is it worthwhile at all?
```

Comment:
> at the call site the arguments are not constexpr, so nothing will change even if you make it constexpr

Probably `constexpr` was not a correct term. I meant making the `newState` known statically, e.g. by moving it to a template parameter:

```cpp
ThreadState SwitchThreadState<ThreadState newState>(ThreadData* threadData) noexcept {
    ...
    isStateSwitchAllowed<newState>(oldState);
}
```

The motivation is to make this check faster. But probably we don't need it and LLVM's inlining + constant propagation will do all the work instead of us. Let me double check that.

Reply:
Right, also since `oldState` is definitely known only at runtime, and you need (1) load, (2) atomic read, I doubt that getting `newState` from a register would have a noticeable impact.